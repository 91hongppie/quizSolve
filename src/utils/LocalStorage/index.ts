import AsyncStorage from '@react-native-async-storage/async-storage';

const LocalStorage = {
  setItem: async (key: string, value: any) => {
    try {
      await AsyncStorage.setItem(key, JSON.stringify(value));
      return await LocalStorage.getItem(key);
    } catch (e) {
      throw new Error('set localStorage failed');
    }
  },

  getItem: async (key: string) => {
    try {
      const keys = await AsyncStorage.getAllKeys();
      if (!keys.includes(key)) {
        return undefined;
      }
      const userInfoStr = await AsyncStorage.getItem(key);
      if (!userInfoStr?.length) {
        return undefined;
      }

      return key !== '@userId' ? JSON.parse(userInfoStr) : userInfoStr;
    } catch (e) {
      throw new Error('get localStorage failed');
    }
  },

  delete: async (key: string) => {
    try {
      AsyncStorage.removeItem(key);
    } catch (e) {
      throw new Error('delete localStorage failed');
    }
  },
  clear: async () => {
    try {
      AsyncStorage.clear();
    } catch (e) {
      throw new Error('clear localStorage failed');
    }
  },
};

export default LocalStorage;
