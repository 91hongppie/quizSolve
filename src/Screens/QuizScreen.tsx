import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {
  Alert,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  SafeAreaView,
  ScrollView,
} from 'react-native';
import {RootStackParamList} from '../../App';
import LocalStorage from '../utils/LocalStorage';
import {useEffect, useState} from 'react';
import {decode} from 'html-entities';
import NavigationBar from '../Components/NavigationBar';

type QuizScreenProps = NativeStackScreenProps<RootStackParamList, 'Quiz'>;

export interface Quiz {
  category: string;
  correct_answer: string;
  difficulty: string;
  incorrect_answers: string[];
  question: string;
  type: string;
}

const QuizScreen = ({navigation, route}: QuizScreenProps) => {
  const [quizCount, setQuizCount] = useState<number>();
  const [quiz, setQuiz] = useState<Quiz>();
  const quizNum = route.params.quizNum;
  const randomNumber = Math.floor(Math.random() * 4);
  const [questions, setQuestions] = useState<[]>();
  const [selectQuestion, setSelectQuestion] = useState<string>();
  const [isSelect, setIsSelect] = useState<boolean>(false);

  useEffect(() => {
    initQuizs();
    startTimer();
  }, []);

  const startTimer = async () => {
    if (quizNum === 0) {
      const startAt = new Date();
      await LocalStorage.setItem('startAt', startAt);
    }
  };
  const initQuizs = async () => {
    const nextQuizs = await LocalStorage.getItem('quizs');
    setQuizCount(nextQuizs.length);
    setQuiz(nextQuizs[quizNum]);
    const nextQuestions = nextQuizs[quizNum].incorrect_answers;
    nextQuestions.splice(randomNumber, 0, nextQuizs[quizNum].correct_answer);
    setQuestions(nextQuestions);
  };

  const goToMain = () => {
    Alert.alert('첫 화면으로 돌아갑니다.', '문제 풀기를 그만하시겠습니까?', [
      {text: '돌아가기', onPress: () => navigation.popToTop()},
      {text: '취소', onPress: () => {}, style: 'destructive'},
    ]);
  };

  const goToNext = async () => {
    if (quizNum === 0) {
      await LocalStorage.setItem('answers', [selectQuestion]);
    } else {
      const answers = await LocalStorage.getItem('answers');
      console.log(answers);
      await LocalStorage.setItem('answers', [...answers, selectQuestion]);
    }
    if (quizCount && quizNum === quizCount - 1) {
      const endAt = new Date();
      await LocalStorage.setItem('endAt', endAt);
      navigation.navigate('Result');
    } else {
      navigation.push('Quiz', {quizNum: quizNum + 1});
    }
  };

  const handleSelectQuestion = (question: string) => {
    if (isSelect) {
      return;
    }
    setSelectQuestion(question);
  };

  const selectAnswer = () => {
    if (!selectQuestion) {
      Alert.alert('정답을 선택해주세요');
      return;
    }
    setIsSelect(true);
  };

  return (
    <SafeAreaView>
      <NavigationBar
        rightFunction={goToNext}
        rightTitle="다음"
        title={`${quizNum + 1} / ${quizNum}`}
        leftFunction={goToMain}
        leftTitle="홈으로"
      />

      <ScrollView style={styles.scrollView}>
        <View style={styles.questionContainer}>
          <Text style={{fontSize: 22}}>{decode(quiz?.category)}</Text>

          <View style={{height: 16}} />
          <View style={{flexDirection: 'row', alignItems: 'center'}}>
            <Text style={{fontSize: 26, color: 'green'}}>
              Quiz {quizNum + 1}
            </Text>
          </View>
          <View style={{height: 10}} />
          <Text style={{fontSize: 26}}>{decode(quiz?.question)}</Text>
          <View style={{height: 16}} />
          <View style={{justifyContent: 'flex-start', gap: 10}}>
            {questions?.map((question, index) => {
              return (
                <TouchableOpacity
                  disabled={isSelect}
                  key={`question_${index}`}
                  style={
                    selectQuestion === question
                      ? styles.selectQuestion
                      : styles.unselectQuestion
                  }
                  onPress={() => {
                    handleSelectQuestion(question);
                  }}>
                  <Text
                    style={
                      selectQuestion === question
                        ? styles.selectQuestionText
                        : styles.unselectQuestionText
                    }>
                    {index + 1}. {decode(question)}
                  </Text>
                </TouchableOpacity>
              );
            })}
          </View>
          <View style={{height: 40}} />
          <View>
            <TouchableOpacity
              style={isSelect ? styles.hide : styles.selectButton}
              onPress={selectAnswer}>
              <Text style={styles.selectButtonText}>선택하기</Text>
            </TouchableOpacity>
            <View style={isSelect ? styles.resultText : styles.hide}>
              <Text
                style={
                  selectQuestion === quiz?.correct_answer
                    ? styles.correctText
                    : styles.failedText
                }>
                {selectQuestion === quiz?.correct_answer
                  ? '정답입니다!'
                  : '틀렸습니다'}
              </Text>
            </View>
          </View>
        </View>
        <View style={{height: 100}} />
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  hide: {
    display: 'none',
  },
  scrollView: {height: '100%'},
  questionContainer: {paddingHorizontal: 10, alignItems: 'center'},
  selectQuestion: {
    borderWidth: 1,
    paddingVertical: 5,
    paddingHorizontal: 10,
    borderRadius: 10,
    borderColor: 'gray',
    backgroundColor: 'green',
  },
  unselectQuestion: {
    borderWidth: 1,
    paddingVertical: 5,
    paddingHorizontal: 10,
    borderRadius: 10,
    borderColor: 'gray',
    backgroundColor: 'transparent',
  },
  selectQuestionText: {
    fontSize: 22,
    color: 'white',
  },
  unselectQuestionText: {
    fontSize: 22,
    color: 'black',
  },
  selectButton: {
    paddingVertical: 10,
    paddingHorizontal: 20,
    backgroundColor: 'green',
    borderRadius: 10,
  },
  selectButtonText: {fontSize: 22, color: 'white'},
  resultText: {
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderRadius: 10,
  },
  correctText: {
    fontSize: 22,
    color: 'green',
  },
  failedText: {
    fontSize: 22,
    color: 'red',
  },
});

export default QuizScreen;
