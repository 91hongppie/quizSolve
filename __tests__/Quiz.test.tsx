import 'react-native';
import React, {ReactElement} from 'react';

// Note: import explicitly to use the types shipped with jest.

// Note: test renderer must be required after react-native.
import QuizScreen from '../src/Screens/QuizScreen';
import {render, waitFor} from '@testing-library/react-native';
import LocalStorage from '../src/utils/LocalStorage';
import '@testing-library/jest-dom';

let props;
let component: ReactElement;

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

function getTempComponent(props: any) {
  return <QuizScreen {...props} />;
}
describe('Quiz', () => {
  props = {route: {params: {quizNum: 0}}};
  component = getTempComponent(props);

  it('renders without crashing', () => {
    const renderResult = render(component);
    const renderedJson = renderResult.toJSON();
    expect(renderedJson).toMatchSnapshot();
    expect(renderedJson).toBeTruthy();
  });

  it('네비게이션 바 왼쪽 버튼 확인', () => {
    const renderResult = render(component);
    expect(renderResult.getByTestId('navibationLeftButton')).toBeDefined();
  });

  it('네비게이션 바 오른쪽 버튼 확인', () => {
    const renderResult = render(component);
    expect(renderResult.getByTestId('navigationRightButton')).toBeDefined();
  });

  it('선택하기 버튼 확인', () => {
    const renderResult = render(component);
    expect(renderResult.getByTestId('answerSelectButton')).toBeDefined();
  });

  it('보기 확인', async () => {
    await LocalStorage.setItem('quizs', [
      {
        category: 'Entertainment: Video Games',
        correct_answer: 'Golden ',
        difficulty: 'medium',
        incorrect_answers: ['Diamond', 'Iron', 'Obsidian'],
        question:
          'Without enchantments, which pickaxe in minecraft mines blocks the quickest.',
        type: 'multiple',
      },
      {
        category: 'History',
        correct_answer: '1983',
        difficulty: 'medium',
        incorrect_answers: ['1934', '1984', '1822'],
        question:
          'What year did Skoal, a smokeless Tobacco company, release their first line of Pouches, known as &quot;Skoal Bandits&quot;?',
        type: 'multiple',
      },
    ]);
    let renderResult = render(component);
    await waitFor(() => {
      expect(renderResult.getByTestId('question_0')).toBeDefined();
      expect(renderResult.getByTestId('question_1')).toBeDefined();
      expect(renderResult.getByTestId('question_2')).toBeDefined();
      expect(renderResult.getByTestId('question_3')).toBeDefined();
    });
    await LocalStorage.delete('quizs');
  });
});
