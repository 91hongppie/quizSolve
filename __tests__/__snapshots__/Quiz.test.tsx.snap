// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz renders without crashing 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "left": 10,
          "opacity": 1,
          "position": "absolute",
        }
      }
      testID="navibationLeftButton"
    >
      <Text
        style={
          {
            "color": "blue",
            "fontSize": 16,
          }
        }
      >
        홈으로
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text>
        1 / 0
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "position": "absolute",
          "right": 10,
        }
      }
      testID="navigationRightButton"
    >
      <Text
        style={
          {
            "color": "green",
            "fontSize": 16,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    style={
      {
        "height": "100%",
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "paddingHorizontal": 10,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 22,
            }
          }
        />
        <View
          style={
            {
              "height": 16,
            }
          }
        />
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              {
                "color": "green",
                "fontSize": 26,
              }
            }
          >
            Quiz 
            1
          </Text>
        </View>
        <View
          style={
            {
              "height": 10,
            }
          }
        />
        <Text
          style={
            {
              "fontSize": 26,
            }
          }
        />
        <View
          style={
            {
              "height": 16,
            }
          }
        />
        <View
          style={
            {
              "gap": 10,
              "justifyContent": "flex-start",
            }
          }
        />
        <View
          style={
            {
              "height": 40,
            }
          }
        />
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "green",
                "borderRadius": 10,
                "opacity": 1,
                "paddingHorizontal": 20,
                "paddingVertical": 10,
              }
            }
            testID="answerSelectButton"
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontSize": 22,
                }
              }
            >
              선택하기
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "height": 100,
          }
        }
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
